added streamlined copy methods to SeqRecord and SeqFeature, and updat…

flaar94 opened the pull request:
…ed the sam parser to use it instead of the much slower built-in recursive deepcopy

- [x] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [x] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [x] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

Closes #4825
Added a .copy method to SeqRecord and SeqFeature classes which provide a faster deep(ish) copy, and changed the sam parser to use this method instead of the built-in python deep copy, following pandas style. Also created a private ._from_validated method in SequenceRecord which is about twice as fast as the standard __init__ on moderate sized data

flaar94 commented:
Draft PR: I'm having some issues with my local black settings

Benchmark Results:
seq_record=Description: <unknown description>
Number of features: 0
Seq('')
Time Elapsed (Shallow Copy (Python Built-in)): 2.741s
Time Elapsed (.copy(deep=False)): 2.851s
Time Elapsed (.copy(deep=True)): 5.767s
Time Elapsed (Python Built-in Deep Copy): 29.755s
seq_record=ID: TestID

Name: TestName
Description: TestDescription
Database cross-references: TestDbxrefs
Number of features: 1
/organism=bombyx
Per letter annotation for: test
Seq('ATGGTGTAAATGGTGTAAATGGTGTAAATGGTGTAAATGGTGTAAATGGTGTAA...TAA')
Time Elapsed (Shallow Copy (Python Built-in)): 2.761s
Time Elapsed (.copy(deep=False)): 2.783s
Time Elapsed (.copy(deep=True)): 9.826s
Time Elapsed (Python Built-in Deep Copy): 60.649s


Sam iterator Benchmark:
9.994804699992528s -> 5.823247199994512s

peterjc commented on Bio/SeqRecord.py:
```diff
@@ -381,6 +382,82 @@ def _set_seq(self, value: Union["Seq", "MutableSeq"]) -> None:
+
+        When deep is True (default) it will make copies of objects the seq, dbxrefs, features, and values of the
+        annotations and letter_annotations dictionaries. However unlike the builtin python copy.deepcopy, this method
+        does not be applied recursively, so if the values of annotations are containers containing mutable objects,
```
does not be -> is not

peterjc commented on Bio/SeqRecord.py:
```diff
@@ -381,6 +382,82 @@ def _set_seq(self, value: Union["Seq", "MutableSeq"]) -> None:
+        When deep is True (default) it will make copies of objects the seq, dbxrefs, features, and values of the
+        annotations and letter_annotations dictionaries. However unlike the builtin python copy.deepcopy, this method
+        does not be applied recursively, so if the values of annotations are containers containing mutable objects,
+        those mutable objects will not be copied. However, this allows it to be significantly faster.
```
So if you did have complex objects, should you use the builtin function? That could be more explicit.

flaar94 replied:
I agree it could be worded better. I was struggling a bit with how to word it. In 99% of cases like that, it should still be fine to use this function, because objects that deep should probably be treated like they’re immutable. You generally don’t want to be doing things like `my_seq_rec.letter_annotations[“some_annotation”][5].append(3)` even if you can. You’d at least want to rewrite like `my_seq_rec.letter_annotations[“some_annotation”][5] = new_list` . Might even be worth asking for it to be at least pseudo-immutable (ie setters turned off), because most of the code is implicitly assuming the sequences one level up are treated as immutable (otherwise lengths could change without checks occurring)

That said, the mutable/immutable distinction is a common gotcha for people not as familiar with Python, and I’d like to avoid circumstances where someone gets an obscure mutability bug and wastes a day failing to find it. This version is roughly following how Pandas defines .copy(deep=True) (except my copy is one level deeper), but I’m not sure if this is best. I also thought about setting the depth parameter as a integer =1 (shallow), 2, 3 (this depth), which would at least give a hint to the user that they might need to go deeper, but that seems like it might be a bit cryptic to the average user

flaar94 commented:
Trying to make empty initialization faster too. I modified _per_letter_annotations to initialize lazily
Python 3.11 benchmarks for 2 million iterations (3.12 is quite a bit faster, but the ratios of times seem the same):
**Master Branch (as of 09-15):**
Time Elapsed (Empty SeqRecord Init): 1.545s
Time Elapsed (Medium SeqRecord Init): 2.803s
Time Elapsed (Highly Annotated Init): 3.291s
Time Elapsed (copy.copy(seq_record) -- medium): 2.656s
Time Elapsed (copy.deepcopy(seq_record) -- medium): 87.327s

Time Elapsed (Sam Parsing (200k lines)): 9.956s
Time Elapsed (SimpleFastaParser (200MB)): 0.702s
Time Elapsed (FastaParser (200MB)): 2.708s

**Lazy initialization:**
Time Elapsed (Empty SeqRecord Init): 0.730s
Time Elapsed (Medium SeqRecord Init): 2.830s
Time Elapsed (Highly Annotated Init): 3.314s
Time Elapsed (Empty SeqRecord _from_validated): 0.591s
Time Elapsed (Medium SeqRecord _from_validated): 1.648s
Time Elapsed (Highly Annotated _from_validated): 1.746s

Time Elapsed (Sam Parsing (200k lines)): 5.353s
Time Elapsed (SimpleFastaParser (200MB)): 0.699s
Time Elapsed (FastaParser (200MB)): 2.134s

Also tested other methods just to make sure there were no regressions, though I didn't try to speed them up (which we should be able to do easily enough by switching from `__init__` to ._from_validated):
**Master Branch (as of 09-15):**
Time Elapsed (Letter Annotation Entry Assignment): 0.701s
Time Elapsed (Full Letter Annotation Assignment): 2.059s
Time Elapsed (Add Moderate + Moderate): 6.506s
Time Elapsed (Add Moderate + Empty): 4.535s
Time Elapsed (Upper Lower): 9.064s
Time Elapsed (Reverse Complement): 10.480s
Time Elapsed (Translate Simple (10% as many iterations)): 4.620s
Time Elapsed (Len): 0.270s
Time Elapsed (iter): 1.681s
Time Elapsed (Count): 1.263s

Time Elapsed (FastqGeneralIterator (200k lines)): 0.244s
Time Elapsed (FastqPhredIterator (200k lines)): 1.526s

**Lazy initialization:**
Time Elapsed (Letter Annotation Entry Assignment): 0.736s
Time Elapsed (Full Letter Annotation Assignment): 1.458s
Time Elapsed (Add Moderate + Moderate): 7.223s
Time Elapsed (Add Moderate + Empty): 4.026s
Time Elapsed (Upper Lower): 9.596s
Time Elapsed (Reverse Complement): 7.112s
Time Elapsed (Translate Simple (10% as many iterations)): 4.604s
Time Elapsed (Len): 0.244s
Time Elapsed (iter): 1.603s
Time Elapsed (Count): 1.261s

Time Elapsed (FastqGeneralIterator (200k lines)): 0.241s
Time Elapsed (FastqPhredIterator (200k lines)): 1.509s

flaar94 commented:
Modified most methods to use ._from_validated(...) when appropriate (and added a check inside of it to default to the class initializer if it gets subclassed to avoid any issues). 
Updated Benchmarks (compared to main branch):
Time Elapsed (Add Moderate + Moderate): 6.506s-> 6.948s
Time Elapsed (Add Moderate + Empty): 4.535s -> 3.796s
Time Elapsed (Upper Lower): 9.064s -> 6.202s
Time Elapsed (Reverse Complement): 10.480 -> 6.952s

_Also modified a couple parsers to use it as a test, along with calling seq_str.encode() which speeds up Seq init_
Time Elapsed (FastaParser (200MB)): 2.708s->1.487s
Time Elapsed (FastqPhredIterator (200k lines)): 1.526s->1.328s

flaar94 commented:
OK, I think that's probably enough things for this pull request. Highlights being that FastaParser and the SAM parser are now nearly twice as fast. Didn't make quite as big of a difference with the QualityIO parser I tested, but it was starting out quite a bit faster to begin with

mdehoon commented on Bio/Align/sam.py:
```diff
@@ -699,7 +699,7 @@ def _read_next_alignment(self, stream):
                 seq = target
                 index = self._target_indices[rname]
-                target = copy.deepcopy(self.targets[index])
+                target = self.targets[index].copy(deep=True)
```
In this case, as `self.targets[index]` does not have complex attributes, is there actually a difference between `copy.deepcopy` and `... .copy(deep=True)`?

flaar94 replied:
The built-in version has to do a lot of extra work searching for all the attributes recursively, and trying to avoid circular references and other kinds of worse-case-scenarios. My main motivation for this method was to help speed up operations in parsers where the types are well known, but I thought it probably was also worth being a public method since a lot of Python module objects have specialized .copy methods

mdehoon replied:
I see. In this case, I would prefer to just create the `SeqRecord` from scratch. The reason is that Biopython always had a problem with software bloat. It is easy to add a function, but removing a function if it does not live up to its expectations is a painful process that takes over a year with deprecation warnings and whatnot. So if there is no absolute need for this public method now, let's not add it.

flaar94 replied:
Totally understandable. I could also change them into a private methods instead. When I did a search, I saw multiple calls to the built-in deepcopy, so there may be a maintenance benefit to code readability and reducing repeated lines. I’ll try to investigate later today to see if there’s enough cases to make it work it

mdehoon replied:
One option is to implement these functions as `__copy__` and `__deepcopy__` (see the Python documentation for these). Then we are not defining new API, but just implementing the API that already exists in Python.

flaar94 replied:
I considered that, but __copy__ is supposed to be a fast shallow copy, which I probably can't do any faster while dealing with _RestrictedDict. I think it would also be best to preserve __deepcopy__ for when we need to be absolutely sure something doesn't share any references, like in tests. I can do this probably about twice as fast by using what we know, but it's probably still too slow to use in hot loops

flaar94 commented:
OK! I have successfully guided my branch through merge conflict purgatory. I removed the copy functions, and replaced the copy in the SAM parser with constructing a new SeqRecord. Since I could customize it, it was slightly faster than my previous code: 4.236s for 200k lines

Note: I think there is something wrong in `QualityIOTestBaseClass` where `compare_record` appeared to be getting stuck in an infinite loop somehow instead of throwing an assert error when comparing quality scores where some were erroneously negative. I didn't figure out exactly how this was happening though

mdehoon commented on Bio/SeqFeature.py:
```diff
@@ -64,6 +64,7 @@
 import functools
 import re
 import warnings
+import copy
```
Do you still need this import?

flaar94 replied:
Oops, no that's just a remnant from me not being careful cleaning up

peterjc commented on Bio/SeqIO/QualityIO.py:
```diff
@@ -538,7 +539,10 @@ def _get_phred_quality(record: SeqRecord) -> Union[list[float], list[int]]:
+        phred_quality = cast(
+            Union[list[float], list[int]], record.letter_annotations["phred_quality"]
+        )
+        return phred_quality
```
Is the type casting an overhead we can avoid another way?

flaar94 replied:
So that `cast` function is from the typing module and it doesn’t actually do anything at runtime ( https://docs.python.org/3/library/typing.html#typing.cast). It’s basically just a way to tell the type checker: “hey, this variable is this type, trust us.” There may be a very slight overhead from doing a function call at all, but it shouldn’t be measurable (I’ll check later). It’s also possible to just turn off the type checker on that line with a comment. I was trying to avoid that if I could help it, but it might be basically the same in this case.

The fundamental challenge here is that the letter_annotations dictionary can have features of any Sequence type, so it’s quite difficult to keep the type checker happy, even if we know a particular kind of annotation is an integer for example

flaar94 replied:
I just removed it and added a type ignore flag to make it look less ugly. I'll try to think if there's any principled way to deal with typing the letter annotations, maybe with typed dictionaries somehow

peterjc replied:
Fixing that nicely is currently beyond my understanding on type annotations. Thanks.

peterjc commented on Bio/SeqIO/SffIO.py:
```diff
@@ -1011,7 +1011,7 @@ def _sff_read_seq_record(self, stream):
         # Dirty trick to speed up this line:
         # record.letter_annotations["phred_quality"] = quals
-        dict.__setitem__(record._per_letter_annotations, "phred_quality", quals)
+        dict.__setitem__(record.letter_annotations, "phred_quality", quals)
```
This should also remove the "dirty trick" comment above. And is this change slowing it down - why not change to use ``record = SeqRecord._from_validated(...)`` as per the FASTQ changes?

flaar94 replied:
Yeah, switching to _from_validated would probably be better. This was an earlier change to mollify the type checker because _per_letter_annotations can now be None if it hasn’t be initialized yet

peterjc commented on Bio/SeqRecord.py:
```diff
@@ -369,17 +371,58 @@ def _set_seq(self, value: Union["Seq", "MutableSeq"]) -> None:
+        """Faster constructor for post-validated data like copies or validated parsed data"""
+
+        if cls is not SeqRecord:
+            # If we subclassed, we'll default to that classes initializer just to be careful
```
classes -> class' (possessive apostrophe)

peterjc reviewed (changes requested):
See minor comments from reading the diff.

This does a lot of nice clean up too (e.g. clearer exceptions when using methods while .seq is None; improving type annotation).

peterjc reviewed (approved):
Are you OK with a squash-and-merge (and let one of us adjust the commit message), or do you want to manually clean up the history ready for a rebase-and-merge?

peterjc commented:
Also I think you deserve your name in the NEWS and CONTRIB files - which can be just the GitHub username if you wish, we have precedent.

flaar94 commented:
> Are you OK with a squash-and-merge (and let one of us adjust the commit message), or do you want to manually clean up the history ready for a rebase-and-merge?

Squash and merge would be fine I think. I'm not sure I have the mental fortitude to clean the git history

flaar94 commented:
OK, changes are now squashed

peterjc commented:
We're going to to need a NEWS entry about this work and the other SeqIO internal changes which have also improved performance.
